### DISCUSSION TITLE
links targeted with `preloadSelectors()` are not preloaded on first page load

### DISCUSSION THREAD

**Discussion by johndwells** [unpoly/unpoly#252] [Q&A]
I have configured Unpoly to preload all links, as well as follow all links (as per https://unpoly.com/handling-everything).

```
up.link.config.preloadSelectors.push(['a[href]']);
up.link.config.followSelectors.push(['a[href]']);
```

However I notice that no links are preloaded on hover, **until a link is followed by Unpoly first**. Meaning, once I clicked a link that Unpoly would follow, _only then_ if I hovered over a link it would be preloaded.

This behaviour doesn't occur if I use `up-preload` attributes on the link; it's only for links that I've configured to preload via `up.link.config.preloadSelectors`.

So I created a codepen to demonstrate, but I couldn't replicate it - my codepen worked fine! After the initial page load, links would automatically preload on hover.

The only difference between the codepen and my webapp is that I'm using npm to build my JS, whereas my codepen was using Unpoly straight from the CDN.

I'm using Vite to build my application files, which uses esbuild during development, and rollup for production builds. In both modes, the built JS introduces this peculiar behavior.

Is this the fault of Vite? Could anyone verify this with an alternative build process, via Webpack, Laravel Mix or similar?

Any other ideas of what I could try to resolve this? Is there something in Unpoly's internals that doesn't attempt preloading until it something is "ready"?

Unpoly Version: 2.2.1
Vite Version: 2.4.4

**Comment by adam12** [comment]
Do you use `async` or `defer` with the `script` tag that loads Unpoly? Where is the `script` tag located? Is it in the `body` or `head`?

It sounds like Unpoly is running too late for the first page load.

**Reply by johndwells** [reply]
@adam12 Ah! Yes with Vite the modern bundle is compiled as an ES6 module, and script tags thus include `type="module"`. This follows the module/nomodule pattern to easily provide modern & legacy builds:

https://philipwalton.com/articles/deploying-es2015-code-in-production-today/

Indeed I can confirm that if I add `type="module"` to the CDN script, the same problem arises. The problem persists regardless of whether the `<script>` tag is in the `<head>` or `<body>`. Also using `async` or `defer` as an alternative to `type="module"` produces the same problem.

I'm relieved to understand the root cause! I wish I could suggest a way to overcome it. Would this be considered a bug?

**Reply by adam12** [reply]
I recreated this with Glitch [here](https://glitch.com/edit/#!/flashy-distinct-hosta). It's working right now, but that's because I cheat and call `up.framework.boot()` again from `scripts.js`. If you comment it out, Unpoly works for everything but doesn't preload links on the page during the first page load.

What I believe is happening is the compiler which sets up the preload is running before the elements are available on the page.

https://github.com/unpoly/unpoly/blob/a5b0fe3ac1d009f771dca9a19afc10288a3686ec/src/unpoly/link.js#L1147-L1151

Perhaps we could wrap that compiler in an `up:app:boot` event listener. 🤔

**Reply by johndwells** [reply]
@adam12 Yep! I can confirm that by calling `up.framework.boot()` from within my app JS, the preload selectors work:

```
// app.js
import 'unpoly';
up.link.config.preloadSelectors.push(['a[href]']);
up.link.config.followSelectors.push(['a[href]']);
up.framework.boot();
```

Speaking of `up:app:boot`, I think that's being triggered "too soon" as well: when imported as a module I cannot successfully bind to the event:

```
// app.js
import 'unpoly';
up.on('up:app:boot', (event) => {
  // this never gets logged
  console.log('booted');
});
```

🤔 indeed...

**Reply by triskweline** [reply]
Note that loading Unpoly with `[defer]` or `[async]` is not supported. `[type=module]` implies `[defer]`. Supporting these cases will require changes in Unpoly.

`up.framework.boot()` is a private function that is automatically called by Unpoly when its script tag gets executed. Calling it manually means that it runs twice, which might set you up for trouble.

**Reply by johndwells** [reply]
@triskweline Thanks for your attention to this issue! I can certainly see where rebooting the application might have unforeseen negative consequences...

But wouldn't it be a good thing to support `[type=module]`? From where I sit, moving towards the module/nomodule pattern continues to gain wide adoption. All modern browsers support module loading, as it's part of es6. Even just to support `defer` or `async` would seem the right move in terms of performance for the end user...

I don't know what changes would be required, and can appreciate if it's more complicated than it might appear on the surface. So I don't ask this lightly but I do think there's a strong case for it.

(Unpoly2 is so wonderful, I've built a number of sites with ~0.6 and couldn't imagine 2.x being that much better, but oh man it is!)

**Reply by adam12** [reply]
> Note that loading Unpoly with `[defer]` or `[async]` is not supported.

I had a feeling this was the case but it was always in the "kind of works, sometimes" territory. I've submitted unpoly/unpoly-site#15 to make this explicit for now.

> But wouldn't it be a good thing to support `[type=module]`

I'm inclined to think this may come, eventually. The project _just_ switched from CoffeeScript to ES6. It's mostly just Henning in his free time where the majority of the current contributions are coming from, so it will depend on his priorities and availability.

**Reply by adam12** [reply]
Actually, I forgot Henning discussed this [just recently](https://github.com/unpoly/unpoly/issues/136#issuecomment-886495652).

**Reply by johndwells** [reply]
Ah ok - I'd be happy to consolidate discussions in that thread.

**Reply by triskweline** [reply]
@adam12 @johndwells Please see [my comment](https://github.com/unpoly/unpoly/issues/136#issuecomment-897106764) in #136.

**Comment by triskweline** [comment]
This will be fixed in the upcoming version 2.3.
